From f4d848720aee3d04aabe50bbbef90ea933a6208c Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Mon, 24 Feb 2025 19:14:45 +0100
Subject: [PATCH 1622/1695] core: do not acknowledge end of interrupt for
 special GIC interrupt IDs

According to the ARM documentation for GICV2/3/4, there is no need to
write to the end of interrupt register for some special IDs. Apply this
recommendation to avoid writing to IO memory in this time sensitive
sequence.

Also distinguish unhandled interrupts with an error log.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 8f4fabc231f487ce541900da5ce8934e55b144b2 core: do not acknowledge end of interrupt for special GIC interrupt IDs]
Change-Id: I0629532f81b1bbf6f8078a8a97e1d744557cc704
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>
[cherry-picked from b711ff7e7d1a78ec0160b5435efed52091ba30b1]
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/440430
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/drivers/gic.c | 27 ++++++++++++++++++++++++++-
 1 file changed, 26 insertions(+), 1 deletion(-)

diff --git a/core/drivers/gic.c b/core/drivers/gic.c
index 444001106..dd7666487 100644
--- a/core/drivers/gic.c
+++ b/core/drivers/gic.c
@@ -655,10 +655,35 @@ static void __maybe_unused gic_native_itr_handler(void)
 	iar = gic_read_iar(gd);
 	id = iar & GICC_IAR_IT_ID_MASK;
 
+	if (id >= 1020 && id <= 1023) {
+		/*
+		 * Special INTIDs
+		 * 1020: Interrupt expected to be handled at SEL1 or SEL2.
+		 *       PE (Processing Element) is either executing at EL3
+		 *       in AArch64 state or in monitor mode in AArch32 state.
+		 *       Reserved on GIC V1 and GIC V2.
+		 * 1021: Interrupt expected to be handled at NSEL1 or NSEL2
+		 *       PE (Processing Element) is either executing at EL3
+		 *       in AArch64 state or in monitor mode in AArch32 state.
+		 *       Reserved on GIC V1 and GIC V2.
+		 * 1022: -(GICv3.3): Interrupt is an NMI
+		 *       -(Legacy): Group 1 interrupt to be signaled to the
+		 *        PE and acknowledged using alias registers. Reserved if
+		 *        interrupt grouping is not supported.
+		 * 1023: No pending interrupt with sufficient priority
+		 *       (spurious) or the highest priority pending interrupt is
+		 *       not appropriate for the current security state or
+		 *       interrupt group.
+		 */
+		DMSG("Special interrupt %"PRIu32, id);
+
+		return;
+	}
+
 	if (id <= gd->max_it)
 		interrupt_call_handlers(&gd->chip, id);
 	else
-		DMSG("ignoring interrupt %" PRIu32, id);
+		EMSG("Unhandled interrupt %"PRIu32, id);
 
 	gic_write_eoir(gd, iar);
 }
-- 
2.47.2

